src/ipc_cache.c

```diff
@@ -0,0 +1,204 @@
+    hash_map_t hash_table;
+    ipc_handle_mapped_cache_eviction_cb_t eviction_cb;
+} ipc_handle_mapped_cache_t;
+
+ipc_handle_mapped_cache_global_t *IPC_MAPPED_CACHE_GLOBAL = NULL;
+
+void umfIpcCacheInit(void) {
+    ipc_handle_mapped_cache_global_t *cache_global =
+        umf_ba_global_alloc(sizeof(*cache_global));
+    if (!cache_global) {
+        return;
+    }
+
+    if (NULL == utils_mutex_init(&(cache_global->cache_lock))) {
+        LOG_ERR("Failed to initialize mutex for the IPC cache");
+        umf_ba_global_free(cache_global);
+        return;
+    }
+
+    cache_global->cache_allocator =
+        umf_ba_create(sizeof(ipc_handle_cache_entry_t));
+    if (!cache_global->cache_allocator) {
+        LOG_ERR("Failed to create IPC cache allocator");
+        umf_ba_global_free(cache_global);
+        return;
+    }
+
+    cache_global->max_size = 0;
+    cache_global->cur_size = 0;
+    cache_global->lru_list = NULL;
+
+    IPC_MAPPED_CACHE_GLOBAL = cache_global;
+}
+
+static size_t getGlobalLruListSize(lru_list_t lru_list) {
+    size_t size = 0;
+    ipc_handle_cache_entry_t *tmp;
+    DL_COUNT(lru_list, tmp, size);
+    return size;
+}
+
+void umfIpcCacheTearDown(void) {
+    ipc_handle_mapped_cache_global_t *cache_global = IPC_MAPPED_CACHE_GLOBAL;
+    IPC_MAPPED_CACHE_GLOBAL = NULL;
+
+    if (!cache_global) {
+        return;
+    }
+
+    assert(cache_global->cur_size == 0);
+    assert(getGlobalLruListSize(cache_global->lru_list) == 0);
+
+    umf_ba_destroy(cache_global->cache_allocator);
+    umf_ba_global_free(cache_global);
+}
+
+ipc_handle_mapped_cache_handle_t umfIpcHandleMappedCacheCreate(
+    ipc_handle_mapped_cache_eviction_cb_t eviction_cb) {
+    ipc_handle_mapped_cache_t *cache = umf_ba_global_alloc(sizeof(*cache));
+
+    if (!cache) {
+        return NULL;
+    }
+
+    cache->global = IPC_MAPPED_CACHE_GLOBAL;
+    cache->hash_table = NULL;
+    cache->eviction_cb = eviction_cb;
+
+    return cache;
+}
+
+void umfIpcHandleMappedCacheDestroy(ipc_handle_mapped_cache_handle_t cache) {
+    ipc_handle_cache_entry_t *entry, *tmp;
+    HASH_ITER(hh, cache->hash_table, entry, tmp) {
+        DL_DELETE(cache->global->lru_list, entry);
+        HASH_DEL(cache->hash_table, entry);
+        cache->global->cur_size -= 1;
+        cache->eviction_cb(&entry->key, &entry->value);
+        umf_ba_free(cache->global->cache_allocator, entry);
+    }
+    HASH_CLEAR(hh, cache->hash_table);
+
+    umf_ba_global_free(cache);
+}
+
+umf_result_t
+umfIpcHandleMappedCacheGet(ipc_handle_mapped_cache_handle_t cache,
+                           const ipc_mapped_handle_cache_key_t *key,
+                           uint64_t handle_id,
+                           ipc_mapped_handle_cache_value_t **retEntry) {
+    ipc_handle_cache_entry_t *entry = NULL;
+    umf_result_t ret = UMF_RESULT_SUCCESS;
+    bool evicted = false;
+    ipc_mapped_handle_cache_value_t evicted_value;
+
+    if (!cache) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
```

Comment:
Done. I am not sure that we need to check `cache->global` because it sets by the `umfIpcHandleMappedCacheCreate` function above. So it is not a user input, but data structure created and intialized by the `umfIpcHandleMappedCacheCreate` function. I prefer to add `assert` in that case.